### DISCUSSION TITLE
[ideas] Warn when name conflicts introduced by static branching

### DISCUSSION THREAD

**Discussion by shiyw** [ropensci/tarchetypes#179] [Ideas]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Currently,  {tarchetypes} does not warn name conflict of names from `values` of `tar_map` and names in `command` of `tar_target`.

```r
# _targets.R
f <- \() cat("Hi")

values <- tibble::tibble(
  f = 1:10
)

list(
  tar_map(
    values = values,
    tar_target(target, f())
  )
)
```

```r
tar_make()
```

```
▶ dispatched target target_6
✖ errored target target_6
▶ completed pipeline [0.198 seconds]
Error:
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    attempt to apply non-function
Last error traceback:
    .handleSimpleError(function (condition)  {     state$error <- build_mess...
    h(simpleError(msg, call))
```

... although `tar_manifest` gives some hints.

```r
tar_manifest()
```
```
# A tibble: 10 × 2
   name      command
   <chr>     <chr>  
 1 target_6  6L()   
 2 target_10 10L()  
 3 target_7  7L()   
 4 target_8  8L()   
 5 target_9  9L()   
 6 target_1  1L()   
 7 target_2  2L()   
 8 target_3  3L()   
 9 target_4  4L()   
10 target_5  5L()   
```

[A repo storing this reprex](https://github.com/shiyw/targets_static_branching_name_conflicts)

**Comment by wlandau** [comment]
Interesting edge case. I still think the general behavior is correct because `tar_map()` operates on static code analysis, and your use of `values` is telling `tar_map()` to treat `f` as a placeholder. The static code analysis and meta programming happen before the pipeline runs, and the former is by design not aware of any objects or functions in memory that the expression may reference.

**Reply by shiyw** [reply]
Yeah, it's unusual usage which no one wish to use that. And current behavior is correct in the sense that commands in `tar_map` run after the global variables are set, so  `values` introduced by `tar_map` should precede the global variables.

My suggestion is should there be a warning in such cases? Since it did take me sometime for me to diagnosis the problem when I unintentionally introduced the conflict. Or, maybe we can mention it in the manual or reference page.

I learned to add some special prefix (like `arg_`) for fields in values to avoid unintended conflicts.
